chemprop/models/ffn.py

```diff
@@ -0,0 +1,355 @@
+                    activation=activation,
+                ),
+                activation,
+            )
+            self.bond_ffn_base = nn.Sequential(
+                build_ffn(
+                    first_linear_dim=2*bond_features_size,
+                    hidden_size=bond_hidden_size,
+                    num_layers=num_layers - 1,
+                    output_size=bond_hidden_size,
+                    dropout=dropout,
+                    activation=activation,
+                ),
+                activation,
+            )
+        else:
+            self.atom_ffn_base = nn.Identity()
+            self.bond_ffn_base = nn.Identity()
+
+        ind = 0
+
+        ffn_list = []
+        for constraint in atom_constraints:
+            ffn_list.append(
+                FFNAtten(
+                    features_size=atom_features_size,
+                    hidden_size=atom_hidden_size,
+                    num_layers=num_layers,
+                    output_size=output_size,
+                    dropout=dropout,
+                    activation=activation,
+                    ffn_base=self.atom_ffn_base,
+                    constraint=constraint,
+                    ffn_type="atom",
+                    shared_ffn=shared_ffn,
+                    weights_ffn_num_layers=weights_ffn_num_layers,
+                )
+            )
+            ind += 1
+
+        for constraint in bond_constraints:
+            ffn_list.append(
+                FFNAtten(
+                    features_size=2*bond_features_size,
+                    hidden_size=bond_hidden_size,
+                    num_layers=num_layers,
+                    output_size=output_size,
+                    dropout=dropout,
+                    activation=activation,
+                    ffn_base=self.bond_ffn_base,
+                    constraint=constraint,
+                    ffn_type="bond",
+                    shared_ffn=shared_ffn,
+                    weights_ffn_num_layers=weights_ffn_num_layers,
+                )
+            )
+            ind += 1
+
+        self.ffn_list = nn.ModuleList(ffn_list)
+
+    def forward(
+        self,
+        input: Tuple[torch.tensor, List, torch.tensor, List, torch.tensor],
+        constraints_batch: List[torch.tensor],
+        bond_types_batch: List[torch.FloatTensor],
+    ) -> List[torch.tensor]:
+        """
+        Runs the :class:`MultiReadout` on input.
+        :param input: A tuple of atomic and bond information of each molecule.
+        :param constraints_batch: A list of PyTorch tensors which applies constraint on atomic/bond properties.
+        :param bond_types_batch: A list of PyTorch tensors storing bond types of each bond determined by RDKit molecules.
+        :return: The output of the :class:`MultiReadout`, a list of PyTorch tensors which ontains atomic/bond properties prediction.
+        """
+        return [ffn(input, constraints_batch[i], bond_types_batch[i]) for i, ffn in enumerate(self.ffn_list)]
```

Comment:
can this be a tensor return type?

Reply:
Both atomic and bond properties can be trained simultaneously, and the length of the returned vector for an atomic property would be different than that for a bond property. So I think it cannot be a tensor type here. It is still possible to get a padded tensor, but it would be too complicated for the following processing (need to additionally define a mask and pass it for the following processes).